publications/radstack/radstack.tex

```diff
@@ -0,0 +1,1148 @@
+into the full details of Kafka/YARN/Samza interactions, but more information is
+available in other literature\cite{2014samza}. We will instead focus on how we
+leverage this framework for processing data for analytic use cases.
+
+On top of Samza infrastructure, we introduce the idea of a “pipeline”, which is
+a grouping for a series of related processing stages where “upstream” stages
+produce messages that are consumed by “downstream” stages. Some of our jobs
+involve operations include renaming data, inserting default values for nulls
+and empty strings, and filtering data. One pipeline may write to many data
+sources in Druid.
+
+Given that Druid does not support joins in queries, we require supporting joins
+at the data processing level. Our approach to do streaming joins is to buffer
+events for a configurable period of time. If an event arrives in the system
+with a join key that exists in the buffer, the join occurs and the joined event
+is transmitted further down the pipeline. If events are substantially delayed
+and do not arrive in the allocated window period, they will not be joined. In
+practice, this generally leads to one “primary” event continuing through the
+pipeline and other secondary events with the same join key getting dropped.
+This means that our stream processing layer is not guaranteed to deliver 100 percent 
+accurate results. Furthermore, even without this restriction, Samza does not
+offer exactly-once processing semantics. Problems in network connectivity or
+node failure can lead to duplicated events. For these reasons, we run a
+separate batch pipeline that generates a more accurate transformation of the
+ingested data.
+
+The final job of our processing pipeline is to deliver data to Druid. For high
+availability, processed events from Samza are transmitted concurrently to two
+real-time nodes. Both nodes receive the same copy of data, and effectively act
+as replicas of each other. The Druid broker does not distinguish between the
+two copies of data, and queries will go to either node. When handoff occurs,
+both real-time nodes race to hand off the segments they’ve created. The segment
+that is pushed into deep storage first will be the one that is used for
+historical querying, and once that segment is loaded on the historical nodes,
+both real-time nodes will drop their versions of the same segment.
+
+\subsection{Batch Processing}
+Our batch processing pipeline is composed a multi-stage
+MapReduce\cite{dean2008mapreduce} pipeline. The first set of jobs mirrors our
+streaming processing pipeline in that it transforms data and joins relevant
+events in preparation for the is data to be loaded into Druid. The second set
+of jobs is responsible for directly creating immutable Druid segments. A final
+set of MapReduce jobs that reads our processed data from the first set and
+indexes this data to create Druid segments. The indexing code for both
```

Comment:
It sounds like the second set of jobs creates segments and then a final, third set of jobs creates segments.